stdlib/LinearAlgebra/src/matmul.jl

```diff
@@ -687,7 +689,7 @@ end
     return _generic_matvecmul!(C, ta, Anew, B, _add)
 end
 
-function _generic_matvecmul!(C::AbstractVector, tA, A::AbstractVecOrMat, B::AbstractVector,
+Base.@constprop :aggressive function _generic_matvecmul!(C::AbstractVector, tA, A::AbstractVecOrMat, B::AbstractVector,
```

Comment:
Unlike the other annotations, this annotation seems to have no effect. Independently from this annotation, I get

```julia
using LinearAlgebra
A = rand(5,5); v= ones(5);
@code_typed mul!(Float32.(v), A, v)
CodeInfo(
1 ─ %1 = invoke LinearAlgebra._generic_matvecmul!(C::Vector{Float32}, 'N'::Char, A::Matrix{Float64}, B::Vector{Float64}, $(QuoteNode(LinearAlgebra.MulAddMul{true, true, Bool, Bool}(true, false)))::LinearAlgebra.MulAddMul{true, true, Bool, Bool})::Vector{Float32}
└──      return %1
) => Vector{Float32}
```

For the other methods, the annotations make that function body of `gemv!` gets "inlined" with known constants. Without the annotation, corresponding `mul!` calls invoke `gemv!`.

Reply:
Yes, this annotation seems to only mark certain branches as dead code once we descend further into `_generic_matvecmul!`. E.g.:
```julia
julia> @descend_code_warntype (() -> mul!(zeros(Int, 3), ones(Int, 3,3), zeros(Int,3), 2, 2))()
[...]
_generic_matvecmul!(C::AbstractVector, tA, A::AbstractVecOrMat, B::AbstractVector, _add::LinearAlgebra.MulAddMul) @ LinearAlgebra ~/julia/usr/share/julia/stdlib/v1.11/LinearAlgebra/src/matmul.jl:692
┌ Warning: Some line information is missing, type-assignment may be incomplete
└ @ Cthulhu ~/.julia/packages/Cthulhu/O1Xhq/src/codeview.jl:144
692 function _generic_matvecmul!(C::Vector{Int64}::AbstractVector, tA::Char, A::Matrix{Int64}::AbstractVecOrMat, B::Vector{Int64}::AbstractVector,
693                             _add::LinearAlgebra.MulAddMul{false, false, Int64, Int64}::MulAddMul = MulAddMul())::Vector{Int64}
694     require_one_based_indexing(C::Vector{Int64}, A::Matrix{Int64}, B::Vector{Int64})
695     @assert tA in ('N', 'T', 'C'::Tuple{Char, Char, Char})
696     mB::Int64 = length(B::Vector{Int64})::Int64
697     mA::Int64, nA::Int64 = lapack_size(tA::Char, A::Matrix{Int64})::Tuple{Int64, Int64}::Int64
698     if (mB::Int64 != nA::Int64)::Bool
699         throw(DimensionMismatch(lazy"matrix A has dimensions ($mA,$nA), vector B has length $mB"))
700     end
701     if (mA::Int64 != length(C::Vector{Int64})::Int64)::Bool
702         throw(DimensionMismatch(lazy"result C has length $(length(C)), needs length $mA"))
703     end
705 
705     Astride::Int64 = size(A::Matrix{Int64}, 1)::Int64
707 
707     @inbounds begin
708     if tA == 'T'  # fastest case
709         if nA == 0
710             for k = 1:mA
711                 _modify!(_add, false, C, k)
712             end
713         else
714             for k = 1:mA
715                 aoffs = (k-1)*Astride
716                 s = zero(A[aoffs + 1]*B[1] + A[aoffs + 1]*B[1])
717                 for i = 1:nA
718                     s += transpose(A[aoffs+i]) * B[i]
719                 end
720                 _modify!(_add, s, C, k)
721             end
722         end
723     elseif tA == 'C'
724         if nA == 0
725             for k = 1:mA
726                 _modify!(_add, false, C, k)
727             end
728         else
729             for k = 1:mA
730                 aoffs = (k-1)*Astride
731                 s = zero(A[aoffs + 1]*B[1] + A[aoffs + 1]*B[1])
732                 for i = 1:nA
733                     s += A[aoffs + i]'B[i]
734                 end
735                 _modify!(_add, s, C, k)
736             end
737         end
738     else # tA == 'N'
739         for i::Int64 = (1:mA::Int64)::Int64::Union{Nothing, Tuple{Int64, Int64}}
740             if !(iszero(_add::LinearAlgebra.MulAddMul{false, false, Int64, Int64}.beta::Int64)::Bool)::Bool
741                 (C::Vector{Int64}[i::Int64]::Int64 *= _add::LinearAlgebra.MulAddMul{false, false, Int64, Int64}.beta::Int64)::Int64
742             elseif mB == 0
743                 C[i] = false
744             else
745                 C[i] = zero(A[i]*B[1] + A[i]*B[1])
746             end
747         end
748         for k::Int64 = (1:mB::Int64)::Int64::Union{Nothing, Tuple{Int64, Int64}}
749             aoffs::Int64 = (((k::Int64-1)::Int64)*Astride::Int64)::Int64
750             b::Int64 = _add::LinearAlgebra.MulAddMul{false, false, Int64, Int64}(B::Vector{Int64}[k::Int64]::Int64)::Int64
751             for i::Int64 = (1:mA::Int64)::Int64::Union{Nothing, Tuple{Int64, Int64}}
752                 (C::Vector{Int64}[i]::Int64 += (A::Matrix{Int64}[(aoffs::Int64 + i)::Int64]::Int64 * b::Int64)::Int64)::Int64
753             end
754         end
755     end
756     end # @inbounds
757     C::Vector{Int64}
758 end
```
This might lead to lesser code being compiled? I don't see a difference in performance, though, so perhaps I'll remove it for now.